design/CommandSubmission.md

```diff
@@ -0,0 +1,23 @@
+
+Since a programmable pass defines the resource binding scope, synchronization rules, fixes the resuorce usage, and exposes a number of specific operations, we encapsulate the encoder of a pass into a separate object, such as `WebGPURenderPassEncoder` and `WebGPUComputePassEncoder`. The pass encoder object can be obtained from a command buffer by calling `beginRenderPass` or `beginComputePass` correspondingly. The command buffer is expected to be in "recording" state, or otherwise a synchronous error is triggered. No operations are expected to be done on the `WebGPUCommandBuffer` if there is an open pass being encoded to it. Calling any methods on the command buffer with an open pass, or submitting it to the command queue, triggers a synchronous error. A pass encoding consists of state setting code and draw/dispatch calls, which are all methods on the corresponding encoder object. In order to close a pass, the user calls `WebGPUProgrammablePassEncoder::endPass`, which returns the owner `WebGPUCommandBuffer` object.
+
+TODO: section about finishing the recording
+
+When a command buffer is recorded, it is in "ready" state. It is valid to transfer this object between web workers. The only operations available on a "ready" command buffer is dropping it and submitting it via `WebGPUCommandQueue::submit`. This method gets a sequence of command buffers and submits them (in the given order) to the GPU driver for execution. There are a few hidden (from the user point of view) stages here before the command buffer actually reaches the GPU.
+
+Once submitted, the command buffer switches to "executing" state. If the WebGPU implementation fails to submit the command buffer due to a problem with recorded content (e.g. exceeding the limit for the instance count in a draw call), it is turned into an internally null object, and the asynchronous error is reported. The feature to re-use command buffers for multiple submissions is still being discussed, and until this is clear, we consider the `WebGPUCommandBuffer` to be moved into submission. Any operations on a command buffer in the "executing" state, other than dropping it (which is what the user is expected to do), would trigger a synchronous error.
+
+If the submission is successful, then at some point in time the GPU will be done processing it. The WebGPU implementation takes the responsibility to detect this moment and gracefully recycle/destroy this command buffer, when it's safe to do so.
+
+## Short version
+
+Command buffers carry sequences of user commands on the CPU side. They can be recorded independently of the GPU, or each other. They go through the following stages:
+
+"recording" -> "ready" -> "executing" -> done
+
+Command buffers are created from and submitted to a command queue. The queue is also used to signal fences, allowing the user to know when the command buffers are done.
```

Comment:
Are command buffers supposed to be submitted in the same order they were created on the queue or not? (I assume the don't but better make sure)